compiler/src/codegen/comp_utils.re

```diff
@@ -133,47 +141,71 @@ let write_universal_exports =
       List.iter(
         item => {
           switch (item) {
-          | TSigValue(id, {val_repr: ReprFunction(args, rets)}) =>
+          | TSigValue(
+              id,
+              {
+                val_repr: ReprFunction(args, rets, direct),
+                val_fullpath: path,
+              },
+            ) =>
             let name = Ident.name(id);
             let exported_name = "GRAIN$EXPORT$" ++ name;
-            let internal_name = Hashtbl.find(exported_names, exported_name);
+            let internal_global_name =
+              Hashtbl.find(exported_names, exported_name);
             let get_closure = () =>
-              Expression.Global_get.make(wasm_mod, internal_name, Type.int32);
+              Expression.Global_get.make(
+                wasm_mod,
+                internal_global_name,
+                Type.int32,
+              );
             let arguments =
               List.mapi(
                 (i, arg) =>
                   Expression.Local_get.make(wasm_mod, i, type_of_repr(arg)),
                 args,
               );
             let arguments = [get_closure(), ...arguments];
-            let call_arg_types =
-              Type.create(
-                Array.of_list(List.map(type_of_repr, [WasmI32, ...args])),
-              );
             let call_result_types =
               Type.create(
                 Array.of_list(
                   List.map(type_of_repr, rets == [] ? [WasmI32] : rets),
                 ),
               );
-            let func_ptr =
-              Expression.Load.make(
-                wasm_mod,
-                4,
-                8,
-                2,
-                Type.int32,
-                get_closure(),
-              );
             let function_call =
-              Expression.Call_indirect.make(
-                wasm_mod,
-                global_function_table,
-                func_ptr,
-                arguments,
-                call_arg_types,
-                call_result_types,
-              );
+              switch (direct) {
+              | Direct(name) =>
+                Expression.Call.make(
+                  wasm_mod,
+                  Hashtbl.find(exported_names, name),
```

Comment:
All these `Hashtbl.find` make me nervous. Are we 100% sure they will always have the values?

Reply:
Yes, and there's no way to recover if it doesn't (compilation failure/bug). We could wrap the whole thing in a `try` with a catch on `Not_found` to say "hey report this if you ever see it" but I didn't feel we really needed that.